### ISSUE TITLE
Can't reproduce results

### ISSUE THREAD

**Issue by xopclabs** [whwjdqls/DEEPTalk#7]
Hey, thanks for open-sourcing your work and spending your time responding to issues here!

I'm not sure if you would be able to help, but I failed to reproduce the results. The thing is, I'm running this repo on pytorch 2.1.0, since 1.X.Y doesn't get shipped for cuda 12.1, which I use. And even though it should be backwards compatible, maybe the problem is simply in this verson mismatch (other problematic libs such as `fairseq` and `funasr` are install according to requirements.txt)  

https://github.com/user-attachments/assets/9376f5e7-8cc6-468d-89c6-7e0723d06228

As you can see, I get little to no lipsync and random (yet fascinatingly expressive, haha) emotions. 

*Another question*: which version of FLAME do you use? I assumed it's 2019, since it what seems Ringnet is using.

Below is `python demo.py` log in case it's useful:
```
['/home/username/code/DEEPTalk/DEEPTalk', '/home/username/.conda/envs/deeptalk/lib/python39.zip', '/home/username/.conda/envs/deeptalk/lib/python3.9', '/home/username/.conda/envs/deeptalk/lib/python3.9/lib-dynload', '/home/username/.conda/envs/deeptalk/lib/python3.9/site-packages', '/home/username/.conda/envs/deeptalk/lib/python3.9/site-packages/setuptools/_vendor', '../', '../']
WARNING:root:Key Conformer already exists in model_classes, re-register
WARNING:root:Key Linear already exists in adaptor_classes, re-register
WARNING:root:Key TransformerDecoder already exists in decoder_classes, re-register
WARNING:root:Key LightweightConvolutionTransformerDecoder already exists in decoder_classes, re-register
WARNING:root:Key LightweightConvolution2DTransformerDecoder already exists in decoder_classes, re-register
WARNING:root:Key DynamicConvolutionTransformerDecoder already exists in decoder_classes, re-register
WARNING:root:Key DynamicConvolution2DTransformerDecoder already exists in decoder_classes, re-register
using device cuda
Loading Models...
DEE config loaded :../DEE/checkpoint/arguments.json
using prob model
funasr version: 1.1.16.
Check update of funasr, and it would cost few times. You may disable it by set `disable_update=True` in AutoModel
New version is available: 1.2.3.
Please use the command "pip install -U funasr" to upgrade.
2025-02-12 21:04:38,211 - modelscope - INFO - PyTorch version 2.1.0+cu121 Found.
2025-02-12 21:04:38,212 - modelscope - INFO - Loading ast index from /home/username/.cache/modelscope/ast_indexer
2025-02-12 21:04:38,443 - modelscope - INFO - Loading done! Current index file version is 1.12.0, with md5 89c9d16224ec323f49a03ce6850c9299 and a total number of 964 components indexed
2025-02-12 21:04:39,718 - modelscope - WARNING - Using the master branch is fragile, please use it with caution!
2025-02-12 21:04:39,718 - modelscope - INFO - Use user-specified model revision: master
Warning, miss key in ckpt: modality_encoders.AUDIO.decoder.blocks.0.0.weight, /home/username/.cache/modelscope/hub/iic/emotion2vec_base_finetuned/emotion2vec_base.pt
Warning, miss key in ckpt: modality_encoders.AUDIO.decoder.blocks.0.0.bias, /home/username/.cache/modelscope/hub/iic/emotion2vec_base_finetuned/emotion2vec_base.pt
Warning, miss key in ckpt: modality_encoders.AUDIO.decoder.blocks.1.0.weight, /home/username/.cache/modelscope/hub/iic/emotion2vec_base_finetuned/emotion2vec_base.pt
Warning, miss key in ckpt: modality_encoders.AUDIO.decoder.blocks.1.0.bias, /home/username/.cache/modelscope/hub/iic/emotion2vec_base_finetuned/emotion2vec_base.pt
Warning, miss key in ckpt: modality_encoders.AUDIO.decoder.blocks.2.0.weight, /home/username/.cache/modelscope/hub/iic/emotion2vec_base_finetuned/emotion2vec_base.pt
Warning, miss key in ckpt: modality_encoders.AUDIO.decoder.blocks.2.0.bias, /home/username/.cache/modelscope/hub/iic/emotion2vec_base_finetuned/emotion2vec_base.pt
Warning, miss key in ckpt: modality_encoders.AUDIO.decoder.blocks.3.0.weight, /home/username/.cache/modelscope/hub/iic/emotion2vec_base_finetuned/emotion2vec_base.pt
Warning, miss key in ckpt: modality_encoders.AUDIO.decoder.blocks.3.0.bias, /home/username/.cache/modelscope/hub/iic/emotion2vec_base_finetuned/emotion2vec_base.pt
Warning, miss key in ckpt: modality_encoders.AUDIO.decoder.proj.weight, /home/username/.cache/modelscope/hub/iic/emotion2vec_base_finetuned/emotion2vec_base.pt
Warning, miss key in ckpt: modality_encoders.AUDIO.decoder.proj.bias, /home/username/.cache/modelscope/hub/iic/emotion2vec_base_finetuned/emotion2vec_base.pt
unfreezing only the first 6 transformer blocks
/home/username/.conda/envs/deeptalk/lib/python3.9/site-packages/huggingface_hub/file_download.py:795: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
/home/username/.conda/envs/deeptalk/lib/python3.9/site-packages/torch/nn/utils/weight_norm.py:30: UserWarning: torch.nn.utils.weight_norm is deprecated in favor of torch.nn.utils.parametrizations.weight_norm.
  warnings.warn("torch.nn.utils.weight_norm is deprecated in favor of torch.nn.utils.parametrizations.weight_norm.")
Some weights of the model checkpoint at facebook/wav2vec2-base-960h were not used when initializing Wav2Vec2ModelResampled: ['lm_head.weight', 'lm_head.bias']
- This IS expected if you are initializing Wav2Vec2ModelResampled from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing Wav2Vec2ModelResampled from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of Wav2Vec2ModelResampled were not initialized from the model checkpoint at facebook/wav2vec2-base-960h and are newly initialized: ['wav2vec2.masked_spec_embed']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
add condition : False
dim factor : 2
**************************************************
DEE
**************************************************
FLINT
**************************************************
Total
audio shape : torch.Size([128000])
audio : torch.Size([1, 128000])
output : torch.Size([1, 200, 53])
saved in ./outputs/params/01_elevenlabs_male_English_5Q0t7uMcjvnagumLfvZi.npy
initializing renderer...
getting flame model and calculating vertices...
creating the FLAME Decoder
vertices shape : torch.Size([200, 5023, 3])
```

**Comment by whwjdqls** [comment]
Hi, 

1. I extracted wav file from the video above and ran the model. This is the result. I don't think the problem is with the version mismatch as, like you mentioned, it should be compatible. Try using FLAME 2020, and if it still result in strange results, provide me the npy file. I will take a look. 

https://github.com/user-attachments/assets/1ddb6472-433e-42e1-a9fc-e0cede9d8030


2. We use FLAME 2020, as we use EMOCA to extract FLAME parameters and EMOCE uses FLAME 2020.

Feel free to ask more questions!

**Comment by xopclabs** [comment]
Thank you! Just got the time to properly reproduce the environment and it was a success, got identical results to yours. Weird, I'll investigate what caused the difference. 

In your paper you mention that you can swap emotion embedding from a different audio and this way control the emotions. Have you tried swapping emotion embedding extracted from expressions instead of audio?

**Comment by whwjdqls** [comment]
I have tried but didn't report the results. 
As DEE is a joint embedding space between expression and audio, it works for both modalities as an emotion control signal.

**Comment by xopclabs** [comment]
Hey there again! I'm trying to test ProbDEE's emotion encoders on RAVDESS but haven't been able to reproduce how emotions cluster using t-SNE. I'm using a self-processed version of RAVDESS for testing.  

My process so far:  
1. **Loading ProbDEE's encoders** – I initialize the audio encoder according to `arguments.json` (finetuned emo2vec), along with the expression encoder and FER.  
2. **Processing data** – I load the audio (resampling to 16kHz, converting to mono) and extract FLAME parameters. Then, I construct the `expression_vector` by concatenating the first 50 expression parameters with 3 jaw rotations. The audio is clipped to `audio_feature_len` (32000 samples, or 2 seconds), while the `expression_vector` is clipped to `expression_feature_len` (60). This might be a bug since 60 corresponds to 2 seconds at 30 fps, but we’re using 25 fps (apparently).  
3. **Inference** – I extract a 128-dimensional feature vector from the `expression_vector` using FER and process both the audio and expression vectors through DEE.  
4. **Iterating over data** – I repeat this process for each data point.  
5. **Dimensionality reduction & visualization** – I concatenate the mean embeddings from audio and expression, then apply t-SNE. Afterward, I split the embeddings back into audio and expression components and create scatter plots.  

Here’s what I get:  
![Image](https://github.com/user-attachments/assets/0dee71c1-3114-4740-839a-6d9be4315aad)  

There’s some separation in the audio embeddings (and to some extent in the expressions), but overall, it’s pretty chaotic. This happens regardless of which 50 expression parameters I use (I’ve tested both halves just in case).  

I was wondering if you could share any FLAME parameters you’ve extracted from RAVDESS or MEAD, as mentioned in the README, so I can cross-check? I’m also considering the possibility of a scaling issue, especially since I used a different model to fit FLAME—though that wouldn't explain why the audio embeddings aren’t clustering nicely either.  Just in case, a sample of my data is available [here](https://drive.google.com/file/d/1Kh6nAA9ulb2X-7M8A2BjGEnfe_UiBYpt/view?usp=sharing). 
```
>>> for k, v in np.load('M017_speech_angry_lvl2_text01_01.npz').items():
...   print(f'{k}: {v.shape}')
...
vertices: (86, 5023, 3)
faces: (9976, 3)
rotation: (86, 3)
translation: (86, 3)
scale: (86, 1, 1)
pose: (86, 3)
neck: (86, 3)
jaw: (86, 3)
eyes: (86, 6)
shape: (1, 300)
expression: (86, 102)
template: (1, 5023, 3)
>>> for k, v in np.load('M017_speech_angry_lvl2_text01_01.npz').items():
...   print(f'{k}: min {np.min(v):.4f}, max {np.max(v):.4f}')
...
vertices: min -0.1843, max 0.1316
faces: min 0.0000, max 5022.0000
rotation: min -0.1557, max -0.0411
translation: min -0.0295, max 2.1853
scale: min 1.0000, max 1.0000
pose: min 0.0000, max 0.0000
neck: min 0.0000, max 0.0000
jaw: min -0.0104, max 0.1203
eyes: min -0.1710, max 0.1102
shape: min -1.3691, max 1.2203
expression: min -1.3118, max 1.2360
template: min -0.1834, max 0.1318
```



On another note, regarding Torch/CUDA versions: I ran my script in both a `torch 1.12.1 + CUDA 11` environment and a `torch 2.1.0 + CUDA 12` setup. The results show that expression mean, std, and audio std embeddings are close within `atol=1e-8`, but some audio mean embeddings are only close within `atol=1e-6` for some reason. It's a bit odd, but the difference seems negligible—so I suspect the real discrepancy lies in the VAE or DEEPTalk decoder itself.

**Comment by xopclabs** [comment]
I've just re-read the thread #4, so I'm downloading preprocessed MEAD now. I'll update this once I try it on the data from there!

**Comment by xopclabs** [comment]
Back with results! Downloaded pre-processed MEAD ran code and got this:
![Image](https://github.com/user-attachments/assets/070a4904-41c1-4a28-8643-27bbaea58136)
For reference, here are the results on my own processing of MEAD (files aren't the same, as I'm sampling random 50 files per `emotion-level` pair each time): ![Image](https://github.com/user-attachments/assets/d0c9494f-b55c-4371-8bac-1dbc98bc47d6)
Both are more or less similar, I'd say. 

What code did you use to get these results in your paper? Maybe I'm missing something only while doing the visualization.

![Image](https://github.com/user-attachments/assets/1048f213-5ca2-419b-b3f2-b0781c8229fe)

**Comment by whwjdqls** [comment]
Hi, 

When I tried visualizing both modalities in the same plot, mine also looked entangled as well. 

Try visualizing them in separate plots, per modality, and if they look entangled too, please share me the images.

**Comment by xopclabs** [comment]
Ah, well, nothing really changes unfortunately! 

![Image](https://github.com/user-attachments/assets/0b58b5a4-a908-4a56-a1bd-e2d7f83f9794)

**Comment by whwjdqls** [comment]
Is this done with the whole MEAD dataset? I only visualized the test set.

If this problem persists, It might be because of the dataset. 

Give me an Email to whwjdqls99@yonsei.ac.kr if you want the exact dataset that I used.

**Comment by xopclabs** [comment]
Hey everyone,

I finally succeeded after some hurdles. As promised in our private conversation, here's a structured summary of our emails (generated with the help of GPT-4.5 and then edited by me), which might help others in reproducing DEEPTalk or understanding the training process of DEE better.

1. Reproducing DEEPTalk Results (Key Points)

- **Expression Encoder Processing**: 
The expression encoder in DEE yielded the best results specifically with the data processed by the DEEPTalk authors themselves. I've experimented with my own processing methods (not EMICA) and even tried the EMICA-processed .hdf5 files provided by Max Planck University, but they consistently produced inferior results. One clear difference was in the range of expression parameters: the authors' data had more extreme values (up to ±9), whereas my own and Max Planck's processing hovered around ±3. Additionally, authors' data exhibited more jitter, which explains the smoothing filter they apply during DEE training.

- **Don't forget Wav2Vec2Processor before DEE.audio_encoder**: 
It's crucial not to skip the wav2vec processing step before passing tensors into the audio encoder. Initially, I missed this step, leading to poor embedding quality. Incorporating the wav2vec processor fixed this, resulting in embeddings very similar to the authors'.

- **FER + Expression Encoder**: 
While the results I eventually obtained matched very closely with the authors', it's important to highlight that the provided expression encoder specifically works well with the authors' own FLAME-based processing of MEAD. If you're processing MEAD differently, you'll likely need to train your own FER model and expression encoder.

2. Insights about training DEE 

- **Emotional consistency loss uses L2 distance instead of reported cosine similarity**: 
Contrary to the cosine similarity mentioned in the paper, the actual implementation employs a simple L2 loss ("prob_method": "mean") between the mean values due to the CSD loss used during DEE training. Using cosine similarity could still be valid, but L2 was effectively sufficient.

- **Why unmentioned in the paper VV_emo_loss was left out**: 
The authors experimented with using the predicted-to-ground-truth expression loss but ultimately omitted it as it had minimal impact. The AV (emotion consistency) loss effectively ensures alignment between audio and expression embeddings, rendering additional VV loss redundant.

- **FER Pre-Training Reasoning**: 
Initially, the DEE training was attempted in an unsupervised manner akin to CLIP. However, due to the small dataset size and limitations with unsupervised contrastive methods (usually requiring very large batch sizes), performance was suboptimal. Thus, FER was introduced as a pre-trained component to provide an explicit emotion prior. Even then, training struggled until the authors incorporated supervised learning directly using the MEAD emotion labels. Also, the code for training DEE isn't updated:
> I just found out that the code in train_prob.py in DEE is not updated yet. If we use [this](https://github.com/whwjdqls/DEEPTalk/blob/ee2fb8ceb962d54b0457cc526fd007ee7873a9a3/DEE/train_prob.py#L50) for an identity matrix for CSD, it would be unsupervised learning (which leads to suboptimal results). Try using [this function](https://github.com/whwjdqls/DEEPTalk/blob/ee2fb8ceb962d54b0457cc526fd007ee7873a9a3/DEE/utils/loss.py#L86) to make masks with emotion labels for supervised learning of DEE.

- **Emotion Consistency Loss Usage in stage2**:
Emotion consistency loss is computationally inexpensive and could theoretically be used from the first stage. Their intuition was that the model first needs to generate accurate expressions, with realism enhancements (such as emotion consistency) applied afterward.

![Image](https://github.com/user-attachments/assets/f6bec941-03eb-4e55-b000-5ab90afd6194)

![Image](https://github.com/user-attachments/assets/04c688dd-31bd-4dc9-821b-8c488a9042a4)

![Image](https://github.com/user-attachments/assets/22f07110-4c0e-4bca-8e1c-bb7ffcd8ca98)

**Comment by whwjdqls** [comment]
Thanks for sharing! Wishing you good luck on your project.

**Comment by Tankaishen** [comment]
Hello, could you share the your process of reproducing? I apply AffectNet dataset and extract 3D flame param of AffectNet and MEAD, train the FER and DEE. Though I followed your instructions above, I can't produce the result that accords with the expectation.
I have questions like below:
1. What important changes need to be applied to the codes or arguments in this repository, as the author mentioned before that the matched matrix  need to be added and the wav2vec2 is needed?Maybe The code and args are not in line with with the results in paper and yours.
2. Did you train your own FER using AffectNet extracted by yourself, or directly use the weights the author provides?
3. If there are other problems that need to be noticed, could you kindly tell me? 
If you would like to, you can reply to me in email cstankaishen@mail.scut.edu.cn or here. Thanks~
